### DISCUSSION TITLE
Game, organizing troops, using the mouse, combining and separating troops of the same type

### DISCUSSION THREAD

**Discussion by LeHerosInconnu** [ihhub/fheroes2#10290] [General]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

In fheroes2, by clicking the central mouse button on a troop, you can fill the free slots of the hero's army with one creature from that troop.
Now, as far as I know, there is no action that allows you to combine all the troops of creatures of the same type into a single troop just by using the mouse.
To do this, you must press and hold the "Alt" key on the keyboard and then left-click on the troop where all the creatures of the same type in the hero's army will be combined.

I also suggest using the central mouse button to combine all troops of creatures of the same type into a single troop.

To do this, as long as there is at least one free slot in the hero's army, the central mouse button performs the same action as at present, i.e., as long as there is a free slot, it is filled with one creature from the starting group.
Now, when there are no more free slots, a click on the central mouse button could have the action of combining all creatures of the same type into a single troop, this troop being the one where the central mouse click is performed.
The current action of combining creatures into a troop using the keyboard can be conserved without any problem.

In fheroes2:
Each click on the central mouse button fills a free slot with one creature from the starting group:

https://github.com/ihhub/fheroes2/assets/43583994/68723795-d17c-446f-86dd-8da69a3a62be

Here is the save file:
[Middle click creature 01.zip](https://github.com/ihhub/fheroes2/files/12830678/Middle.click.creature.01.zip)

### Describe the possible solution

Allow the central mouse button to combine all troops of creatures of the same type into a single troop.

To do this, as long as there is at least one free slot in the hero's army, the central mouse button performs the same action as at present, i.e., as long as there is a free slot, it is filled with one creature from the starting group.
Now, when there are no more free slots, a click on the central mouse button could have the action of combining all creatures of the same type into a single troop, this troop being the one where the central mouse click is performed.
The current action of combining creatures into a troop using the keyboard can be conserved without any problem.

### Additional info

Version 1.0.8 7767

**Comment by oleg-derevenetz** [comment]
> To do this, as long as there is at least one free slot in the hero's army, the central mouse button performs the same action as at present, i.e., as long as there is a free slot, it is filled with one creature from the starting group.
Now, when there are no more free slots, a click on the central mouse button could have the action of combining all creatures of the same type into a single troop, this troop being the one where the central mouse click is performed.

I don't think it's a good idea. Many people (including me, sometimes) just keep clicking "while it works" and can easily make extra clicks. Now it's absolutely harmless - with extra clicks, nothing will happen. However, if this proposal is implemented, one extra click - and you will need to start over.

**Comment by Branikolog** [comment]
Hi, @LeHerosInconnu and @oleg-derevenetz 
What if use "hold click" method, which was implemented recently for adventure map "continue movement" button?
So just hold central button and after ~700-800ms all creatures will be combined into a single stack, what do you think?

**Comment by oleg-derevenetz** [comment]
> What if use "hold click" method, which was implemented recently for adventure map "continue movement" button?
So just hold central button and after ~700-800ms all creatures will be combined into a single stack, what do you think?

If you have a mouse with the middle button, then you most probably have the keyboard as well, so there is no problem with holding the Alt or Ctrl button. Therefore, the constructive side of the proposal is unclear to me - what exactly can be done that could not have been done before in similar conditions.

**Comment by LeHerosInconnu** [comment]
Hello @oleg-derevenetz and @Branikolog,

> > To do this, as long as there is at least one free slot in the hero's army, the central mouse button performs the same action as at present, i.e., as long as there is a free slot, it is filled with one creature from the starting group.
> > Now, when there are no more free slots, a click on the central mouse button could have the action of combining all creatures of the same type into a single troop, this troop being the one where the central mouse click is performed.
> 
> I don't think it's a good idea. Many people (including me, sometimes) just keep clicking "while it works" and can easily make extra clicks. Now it's absolutely harmless - with extra clicks, nothing will happen. However, if this proposal is implemented, one extra click - and you will need to start over.

At worst, it would just be a few more mouse clicks for the unlimited mouse-click addicts.
And after a while of use, the muscle memory will naturally kick in to make just the right number of clicks.

> Hi, @LeHerosInconnu and @oleg-derevenetz What if use "hold click" method, which was implemented recently for adventure map "continue movement" button? So just hold central button and after ~700-800ms all creatures will be combined into a single stack, what do you think?

This could lead to erratic operation.
A user could press the central mouse button (which, let's not forget, is a little harder to press than the left and right buttons) for a longer or shorter time, resulting in an undesired action.
This would certainly be considered and reported as a bug.
It would also require greater concentration and memory on the part of the user, who would have to remember to apply a longer pressure to reunite creatures than to separate them.

> > What if use "hold click" method, which was implemented recently for adventure map "continue movement" button?
> > So just hold central button and after ~700-800ms all creatures will be combined into a single stack, what do you think?
> 
> If you have a mouse with the middle button, then you most probably have the keyboard as well, so there is no problem with holding the Alt or Ctrl button. Therefore, the constructive side of the proposal is unclear to me - what exactly can be done that could not have been done before in similar conditions.

The aim is to use only the mouse and not the keyboard at all.

**Comment by Branikolog** [comment]
@LeHerosInconnu 
So using central button for splitting a single troop is ok, but holding for a few moments - undesired action? :) If player accidentally fails and holds button for less, than required - the only thing happens - a single troop will be splitted. But on the second attempt player will combine all together, so there's no worry about accidentally made stacks, as the initial goal was to combine all together.

@oleg-derevenetz 

>If you have a mouse with the middle button, then you most probably have the keyboard as well, so there is no problem with holding the Alt or Ctrl button. Therefore, the constructive side of the proposal is unclear to me - what exactly can be done that could not have been done before in similar conditions.

We have already the interaction logic by clicking on central wheel/button (split 1 single troop). So why not expand the possibilities for those people, who are playing laying on the bad without keyboard with just a mouse? :) 
It's just some sort of improving experience, as it doesn't break any existing logic or interrupt using it.

**Comment by oleg-derevenetz** [comment]
> We have already the interaction logic by clicking on central wheel/button (split 1 single troop). So why not expand the possibilities for those people, who are playing laying on the bad without keyboard with just a mouse? :)
It's just some sort of improving experience, as it doesn't break any existing logic or interrupt using it.

From my point of view, it looks like this improvement doesn't bring anything new. It will be used only on a PC by 3.5 people who know about it, and who already have the opportunity to do the same thing, just in a different way, but this will not work on touchscreen devices and on game consoles.

**Comment by Branikolog** [comment]
@oleg-derevenetz 
The same as current click on the wheel function to split 1 stack, which could be used on PCs with mouse manipulator only.
Actually, that's what the issue is all about.

**Comment by LeHerosInconnu** [comment]
@oleg-derevenetz.

> From my point of view, it looks like this improvement doesn't bring anything new. It will be used only on a PC by 3.5 people who know about it, and who already have the opportunity to do the same thing, just in a different way, but this will not work on touchscreen devices and on game consoles.

Your arguments are neither substantiated nor constructive.
If it's possible to improve the use of the game, even if it's not for all types of hardware, why not do it?
We're also open to your proposals for universal operation on all types of hardware.

**Comment by oleg-derevenetz** [comment]
> Your arguments are neither substantiated nor constructive.
If it's possible to improve the use of the game, even if it's not for all types of hardware, why not do it?

I don't see any "improvement" here. What used to be possible to do only on a PC can still be done only on a PC, just in another way, which still needs to be guessed, because this method is not explained/suggested in the game UI in any way.

> We're also open to your proposals for universal operation on all types of hardware.

The absence of such proposals does not mean that we need to implement some half-baked PC-only behavior, especially since there is already a way to do the same on PC. There is no consensus even on the essence of the proposal, because using the same action to perform opposite tasks (sometimes splitting, and sometimes combining) obviously inconvenient, because one slip-up and you did something completely different.

**Comment by LeHerosInconnu** [comment]
@oleg-derevenetz.

> > Your arguments are neither substantiated nor constructive.
> > If it's possible to improve the use of the game, even if it's not for all types of hardware, why not do it?
> 
> I don't see any "improvement" here.

The improvement is that the action can be performed using only and exclusively the mouse.

> > We're also open to your proposals for universal operation on all types of hardware.
> 
> The absence of such proposals does not mean that we need to implement some half-baked PC-only behavior, especially since there is already a way to do the same on PC.

The improvement will allow the action to be performed using only and exclusively the mouse, which is not currently the case.

> There is no consensus even on the essence of the proposal, because using the same action to perform opposite tasks (sometimes splitting, and sometimes combining) obviously inconvenient, because one slip-up and you did something completely different.

It's a way of working like any other, and splitting and combining creatures are not actions as far apart as that in the principle.

**Comment by ihhub** [comment]
We are trying to add a vital game mechanic using a limited way. Not many people use mice with clickable wheels (hello, Magic Mouse) or they even use just touchpads. The game is designed to be played by a keyboard with a mouse. Touchscreen players are currently limited in the gameplay and with this idea they would be even more limited as it is impossible to simulate (without some crazy implementation) a middle mouse click.

Moreover, such feature must be reflected in the list of hot keys and as of now it is not possible to implement without changing a lot of code.

If we go with implementing army merging by some method it should be accessible by everyone first and only then by chosen ones (like me) with more options.

Also, I suggest to refrain proposing long press events till we gather feedback from players about this for the hero movement mechanics.

A bonus idea to think: we can implement proper a drag-and-drop mechanics within the army window. I have a prototype code for this.

**Comment by ihhub** [comment]
Moving this topic to the Discussion section as we haven't found agreement on this yet.
